### PR TITLE
BUG: 35977 Adding regex support for ExtensionBlock replace method

### LINKED ISSUE #35977: BUG: replacement works for object but not string dtype
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample

```python
import pandas as pd
import numpy as np

a = pd.DataFrame({'a': ['a', 'b', 'c'], 'b': ['d', '', '']}, dtype='object')
b = pd.DataFrame({'a': ['a', 'b', 'c'], 'b': ['d', '', '']}, dtype='string')

print(a)
a.replace(r'^\s*$', pd.NA, regex=True, inplace=True)
print(a)

print(b)
b.replace(r'^\s*$', pd.NA, regex=True, inplace=True)
print(b)
```

#### Problem description

replace(r'^\s*$', pd.NA, regex=True, inplace=True) works on object dtype, but not on stringdtype

#### Expected Output

Same on both replacement

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here leaving a blank line after the details tag]

</details>

### PR DESCRIPTION
**Pull request by cgangwar11**
- [x] closes #35977 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Overrides replace method to add support of regex in ExtensionBlock . If input to_replace is a regex we convert data type of block to Object and calls object replace method which supports regex otherwise we call replace method of super class Block.

```ipython
In [2]: b.replace(r'^\s*$', pd.NA, regex=True, inplace=True)
   ...: print(b)
   a     b
0  a     d
1  b  <NA>
2  c  <NA>

In [3]: b.replace(r'^\s*$', pd.NA, regex=True, inplace=False)
Out[3]: 
   a     b
0  a     d
1  b  <NA>
2  c  <NA>


```

### CODE REVIEW

File: pandas/tests/arrays/categorical/test_replace.py [hunk 1]
```diff
@@ -49,3 +49,51 @@ def test_replace(to_replace, value, expected, flip_categories):
...
+
+@pytest.mark.parametrize(
+    "to_replace,value,input_data,expected_data",
+    [
+        (r"^\s*$", pd.NA, ["d", "ee", "f", ""], ["d", "ee", "f", pd.NA]),
+        (r"e{2}", "replace", ["d", "ee", "f", ""], ["d", "replace", "f", ""]),
+        (r"f", "replace", ["d", "ee", "f", ""], ["d", "ee", "replace", ""]),
+    ],
+)
+def test_replace_regex(to_replace, value, input_data, expected_data):
```
Comment (dsaxton) [inline pandas/tests/arrays/categorical/test_replace.py:78]: If you parametrize this over inplace you can deduplicate these tests a lot

Comment (cgangwar11) [reply pandas/tests/arrays/categorical/test_replace.py]: @dsaxton addressed this pr comment. Summary created two test cases to verify changes with regex and without regex.

**Comment by jreback** [comment]
pls merge master as we just merged a change in how replace works in internals.

**Comment by cgangwar11** [comment]
@jreback synced with upstream master also verified that this bug is still present in upstream master

**Comment by simonjayhawkins** [comment]
@cgangwar11

Conflicting files
pandas/tests/arrays/categorical/test_replace.py

File: pandas/core/internals/blocks.py [hunk 2]
```diff
@@ -1961,6 +1961,33 @@ def _unstack(self, unstacker, fill_value, new_placement):
...
+    ):
+        """
+        replace the to_replace value with value, regex is not supported by super class
+        when regex is required ObjectBlock replace method is called
+        """
+        inplace = validate_bool_kwarg(inplace, "inplace")
+        regex = validate_bool_kwarg(regex, "regex")
+        if regex:
+            dtype = self.values.dtype
+            blocks = self.astype(object)
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:1980]: can you give this a non-plural name

File: doc/source/whatsnew/v1.2.0.rst [hunk 3]
```diff
@@ -462,9 +462,11 @@ Sparse
 ExtensionArray
 ^^^^^^^^^^^^^^
 
+
 - Fixed Bug where :class:`DataFrame` column set to scalar extension type via a dict instantion was considered an object type rather than the extension type (:issue:`35965`)
 - Fixed bug where ``astype()`` with equal dtype and ``copy=False`` would return a new object (:issue:`284881`)
 - Fixed bug when applying a NumPy ufunc with multiple outputs to a :class:`pandas.arrays.IntegerArray` returning None (:issue:`36913`)
+- Fixed bug in :meth:`Dataframe.replace` not working for ``regex=True`` and ``dtype='string'`` Now working as expected (:issue:`35977`)
```
Comment (arw2019) [inline doc/source/whatsnew/v1.2.0.rst:469]: Please change this to:
```
- Fixed bug in :meth:`DataFrame.replace` now returns correct result for `regex=True`` with ``string`` dtype (:issue:`35977`)
```

File: pandas/tests/arrays/categorical/test_replace.py [hunk 4]
```diff
@@ -45,3 +45,41 @@ def test_replace(to_replace, value, expected, flip_categories):
...
+    "to_replace,value,input_data,expected_data",
+    [
+        ("", pd.NA, ["d", "ee", "f", ""], ["d", "ee", "f", pd.NA]),
+        ("ee", "replace", ["d", "ee", "f", ""], ["d", "replace", "f", ""]),
+        ("f", "replace", ["d", "ee", "f", ""], ["d", "ee", "replace", ""]),
+    ],
+)
+def test_replace_string(to_replace, value, input_data, expected_data):
+    # GH35977
+    input_df = pd.DataFrame({"col1": input_data}, dtype="string")
```
Comment (arw2019) [inline pandas/tests/arrays/categorical/test_replace.py:82]: Call this `df`

File: pandas/tests/arrays/categorical/test_replace.py [hunk 5]
```diff
@@ -45,3 +45,41 @@ def test_replace(to_replace, value, expected, flip_categories):
...
+    expected_df = pd.DataFrame({"col1": expected_data}, dtype="string")
+    input_df = input_df.replace(to_replace, value, inplace=False, regex=False)
```
Comment (arw2019) [inline pandas/tests/arrays/categorical/test_replace.py:84]: Call this `result`

File: pandas/tests/arrays/categorical/test_replace.py [hunk 6]
```diff
@@ -45,3 +45,41 @@ def test_replace(to_replace, value, expected, flip_categories):
...
+    [
+        ("", pd.NA, ["d", "ee", "f", ""], ["d", "ee", "f", pd.NA]),
+        ("ee", "replace", ["d", "ee", "f", ""], ["d", "replace", "f", ""]),
+        ("f", "replace", ["d", "ee", "f", ""], ["d", "ee", "replace", ""]),
+    ],
+)
+def test_replace_string(to_replace, value, input_data, expected_data):
+    # GH35977
+    input_df = pd.DataFrame({"col1": input_data}, dtype="string")
+    expected_df = pd.DataFrame({"col1": expected_data}, dtype="string")
```
Comment (arw2019) [inline pandas/tests/arrays/categorical/test_replace.py:83]: Call this `expected`

File: pandas/tests/arrays/categorical/test_replace.py [hunk 7]
```diff
@@ -45,3 +45,41 @@ def test_replace(to_replace, value, expected, flip_categories):
...
+
+@pytest.mark.parametrize(
+    "to_replace,value,input_data,expected_data",
+    [
+        ("", pd.NA, ["d", "ee", "f", ""], ["d", "ee", "f", pd.NA]),
+        ("ee", "replace", ["d", "ee", "f", ""], ["d", "replace", "f", ""]),
+        ("f", "replace", ["d", "ee", "f", ""], ["d", "ee", "replace", ""]),
+    ],
+)
+def test_replace_string(to_replace, value, input_data, expected_data):
```
Comment (arw2019) [inline pandas/tests/arrays/categorical/test_replace.py:80]: Use the idiom:
```
result = ...
expected = ...
tm.assert_frame_equal(result, frame)
```

File: pandas/tests/arrays/categorical/test_replace.py [hunk 8]
```diff
@@ -45,3 +45,41 @@ def test_replace(to_replace, value, expected, flip_categories):
...
+        (r"e{2}", "replace", ["d", "ee", "f", ""], ["d", "replace", "f", ""], True),
+        (r"f", "replace", ["d", "ee", "f", ""], ["d", "ee", "replace", ""], True),
+    ],
+)
+def test_replace_regex(to_replace, value, input_data, expected_data, inplace):
+    # GH35977
+    input_df = pd.DataFrame({"col1": input_data}, dtype="string")
+    expected_df = pd.DataFrame({"col1": expected_data}, dtype="string")
+    converted = input_df.replace(to_replace, value, inplace=inplace, regex=True)
+    if inplace:
```
Comment (arw2019) [inline pandas/tests/arrays/categorical/test_replace.py:66]: do 
```
result = df_replaced if inplace else df
```

File: pandas/tests/arrays/categorical/test_replace.py [hunk 9]
```diff
@@ -45,3 +45,41 @@ def test_replace(to_replace, value, expected, flip_categories):
...
+        (r"e{2}", "replace", ["d", "ee", "f", ""], ["d", "replace", "f", ""], False),
+        (r"f", "replace", ["d", "ee", "f", ""], ["d", "ee", "replace", ""], False),
+        (r"^\s*$", pd.NA, ["d", "ee", "f", ""], ["d", "ee", "f", pd.NA], True),
+        (r"e{2}", "replace", ["d", "ee", "f", ""], ["d", "replace", "f", ""], True),
+        (r"f", "replace", ["d", "ee", "f", ""], ["d", "ee", "replace", ""], True),
+    ],
+)
+def test_replace_regex(to_replace, value, input_data, expected_data, inplace):
+    # GH35977
+    input_df = pd.DataFrame({"col1": input_data}, dtype="string")
```
Comment (arw2019) [inline pandas/tests/arrays/categorical/test_replace.py:63]: Call this `df`

File: pandas/tests/arrays/categorical/test_replace.py [hunk 10]
```diff
@@ -45,3 +45,41 @@ def test_replace(to_replace, value, expected, flip_categories):
...
+    expected_df = pd.DataFrame({"col1": expected_data}, dtype="string")
```
Comment (arw2019) [inline pandas/tests/arrays/categorical/test_replace.py:64]: Call this `expected`

File: pandas/tests/arrays/categorical/test_replace.py [hunk 11]
```diff
@@ -45,3 +45,41 @@ def test_replace(to_replace, value, expected, flip_categories):
...
+    converted = input_df.replace(to_replace, value, inplace=inplace, regex=True)
```
Comment (arw2019) [inline pandas/tests/arrays/categorical/test_replace.py:65]: Call this `df_replaced`

File: pandas/tests/arrays/categorical/test_replace.py [hunk 12]
```diff
@@ -45,3 +45,41 @@ def test_replace(to_replace, value, expected, flip_categories):
...
+        (r"^\s*$", pd.NA, ["d", "ee", "f", ""], ["d", "ee", "f", pd.NA], False),
+        (r"e{2}", "replace", ["d", "ee", "f", ""], ["d", "replace", "f", ""], False),
+        (r"f", "replace", ["d", "ee", "f", ""], ["d", "ee", "replace", ""], False),
+        (r"^\s*$", pd.NA, ["d", "ee", "f", ""], ["d", "ee", "f", pd.NA], True),
+        (r"e{2}", "replace", ["d", "ee", "f", ""], ["d", "replace", "f", ""], True),
+        (r"f", "replace", ["d", "ee", "f", ""], ["d", "ee", "replace", ""], True),
+    ],
+)
+def test_replace_regex(to_replace, value, input_data, expected_data, inplace):
+    # GH35977
```
Comment (arw2019) [inline pandas/tests/arrays/categorical/test_replace.py:62]: Use the idiom:
```
result = ...
expected = ...
tm.assert_frame_equal(result, frame)
```

File: pandas/core/internals/blocks.py (hunk 2 above)
Comment (arw2019) [reply pandas/core/internals/blocks.py:1980]: @cgangwar11 can you address?

**Review by arw2019 — CHANGES_REQUESTED** [review]

**Comment by arw2019** [comment]
Also will need to merge master

File: pandas/tests/arrays/categorical/test_replace.py [hunk 13]
```diff
@@ -45,3 +45,41 @@ def test_replace(to_replace, value, expected, flip_categories):
...
+    [
+        (r"^\s*$", pd.NA, ["d", "ee", "f", ""], ["d", "ee", "f", pd.NA], False),
+        (r"e{2}", "replace", ["d", "ee", "f", ""], ["d", "replace", "f", ""], False),
+        (r"f", "replace", ["d", "ee", "f", ""], ["d", "ee", "replace", ""], False),
+        (r"^\s*$", pd.NA, ["d", "ee", "f", ""], ["d", "ee", "f", pd.NA], True),
+        (r"e{2}", "replace", ["d", "ee", "f", ""], ["d", "replace", "f", ""], True),
+        (r"f", "replace", ["d", "ee", "f", ""], ["d", "ee", "replace", ""], True),
+    ],
+)
+def test_replace_regex(to_replace, value, input_data, expected_data, inplace):
```
Comment (jbrockmendel) [inline pandas/tests/arrays/categorical/test_replace.py:61]: if this is testing DataFrame.replace, these belong in tests.frame.methods.test_replace

File: pandas/tests/frame/methods/test_replace.py [hunk 14]
```diff
@@ -1632,3 +1632,37 @@ def test_replace_unicode(self):
...
+
+    @pytest.mark.parametrize(
+        "to_replace,value,input_data,expected_data",
+        [
+            ("", pd.NA, ["d", "ee", "f", ""], ["d", "ee", "f", pd.NA]),
+            ("ee", "replace", ["d", "ee", "f", ""], ["d", "replace", "f", ""]),
+            ("f", "replace", ["d", "ee", "f", ""], ["d", "ee", "replace", ""]),
+        ],
+    )
+    def test_replace_string(self, to_replace, value, input_data, expected_data):
```
Comment (jbrockmendel) [inline pandas/tests/frame/methods/test_replace.py:1663]: can you parametrize these using frame_or_series

File: pandas/core/internals/blocks.py [hunk 15]
```diff
@@ -2027,6 +2027,32 @@ def _unstack(self, unstacker, fill_value, new_placement):
...
+        if regex:
+            dtype = self.values.dtype
+            block = self.astype(object)
+            if not inplace:
+                return [
+                    b.astype(dtype)
+                    for b in block.replace(to_replace, value, inplace, regex)
+                ]
+            block.replace(to_replace, value, inplace, regex)
+            return block.astype(dtype)
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:2052]: We recently implement Block._replace_regex.  can you see if we can avoid the object-dtype cast (ideally even this whole method) by using that?

Comment (cgangwar11) [reply pandas/core/internals/blocks.py:2054]: replace_regex is using np.vectorize function internally which gives an error when using not object dtype hence using replace_regex after typecasting.
f = np.vectorize(re_replacer, otypes=[values.dtype])

File: pandas/tests/frame/methods/test_replace.py (hunk 14 above)
Comment (arw2019) [reply pandas/tests/frame/methods/test_replace.py]: was this addressed?

**Review by arw2019 — COMMENTED** [review]
@cgangwar11 what's the status of this? it sounds like you addressed most (possibly all) comments so if you merge master and move the whatsnew to 1.3 we'll take another look

One question I have is whether this affects the `ExtensionDtype` `Series.replace` case? If yes we'd like to test with series

Another question is whether we have asvs for this? (If not that doesn't have to be part of this PR IMO)

**Comment by jreback** [comment]
closing as stale. if you want to continue, pls ping and can re-open.

**Pull request closed without merging**